### PR TITLE
[WebAssembly] Protect memory.fill and memory.copy from zero-length ranges.

### PR DESCRIPTION
**Pull request by sunfishcode**
WebAssembly's `memory.fill` and `memory.copy` instructions trap if the pointers are out of bounds, even if the length is zero. This is different from LLVM, which expects that it can call `memcpy` on arbitrary invalid pointers if the length is zero. To avoid spurious traps, branch around `memory.fill` and `memory.copy` when the length is zero.

### CODE REVIEW

File: llvm/lib/Target/WebAssembly/WebAssemblyInstrBulkMemory.td [hunk 1]
```diff
@@ -21,16 +21,33 @@ multiclass BULK_I<dag oops_r, dag iops_r, dag oops_s, dag iops_s,
 }
 
 // Bespoke types and nodes for bulk memory ops
+
+// memory.copy (may trap on empty ranges)
+def wasm_memory_copy_t : SDTypeProfile<0, 5,
+  [SDTCisInt<0>, SDTCisInt<1>, SDTCisPtrTy<2>, SDTCisPtrTy<3>, SDTCisInt<4>]
+>;
```
Comment (aheejin) [inline llvm/lib/Target/WebAssembly/WebAssemblyInstrBulkMemory.td:28]: This seems identical to the existing `wasm_memcpy_t`.. Can we reuse it in both places? The same for `wasm_memory_fill_t`.

File: llvm/lib/Target/WebAssembly/WebAssemblyInstrBulkMemory.td [hunk 2]
```diff
@@ -51,25 +68,83 @@ defm DATA_DROP :
...
          "data.drop\t$seg", "data.drop\t$seg", 0x09>;
 
+}
+
+defm : BulkMemoryOps<I32, "32">;
+defm : BulkMemoryOps<I64, "64">;
+
+// Define copy/fill manually instead of using the `BulkMemoryOps` multiclass
+// because when a multiclass defines opcodes, it gives them anonymous names
+// and we need opcodes with names so that we can handle them with custom code.
```
Comment (aheejin) [inline llvm/lib/Target/WebAssembly/WebAssemblyInstrBulkMemory.td:78]: Not sure what this means. We use all opcodes in code by `WebAssembly::INSTRNAME` and most of them all inherit from some multiclass..? For example, this is from `LowerFPToInt`:
https://github.com/llvm/llvm-project/blob/27f66c9a84bc72acc2536013e08ea0a11161e81b/llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp#L481-L487
All instructions here inherit from some multiclass. For example, `ABS_F32` inherits from the multiclass `UNARY_FP`, and we can use their names fine:
https://github.com/llvm/llvm-project/blob/267be4a7f4ac69cfd1bec5223554bbe400c5636c/llvm/lib/Target/WebAssembly/WebAssemblyInstrFloat.td#L55

Am I missing something?

Comment (sunfishcode) [reply llvm/lib/Target/WebAssembly/WebAssemblyInstrBulkMemory.td]: Ah, it appears to be due to the `defm : BulkMemoryOps<...>`, which makes the opcodes "anonymous", and tablegen names their opcodes as `anonymous_9000MEMORY_COPY_A64` and `anonymous_9000MEMORY_COPY_A64` and `anonymous_9000MEMORY_FILL_A64`. If I change it to `defm : MEMORY_ : BulkMemoryOps<...>` and then remove the `MEMORY_` prefix from all the names, then they come out as just `MEMORY_COPY_A64` and `MEMORY_FILL_A64`.

I've now updated the code to use multiclasses for all instructions that have 32/64 variants.

File: llvm/lib/Target/WebAssembly/WebAssemblyInstrBulkMemory.td (hunk 1 above)
Comment (sunfishcode) [reply llvm/lib/Target/WebAssembly/WebAssemblyInstrBulkMemory.td]: Done

**Review by aheejin — COMMENTED** [review]
The tests don't seem to be working now

File: llvm/lib/Target/WebAssembly/WebAssemblyInstrBulkMemory.td [hunk 3]
```diff
@@ -21,22 +21,37 @@ multiclass BULK_I<dag oops_r, dag iops_r, dag oops_s, dag iops_s,
 }
 
 // Bespoke types and nodes for bulk memory ops
-def wasm_memcpy_t : SDTypeProfile<0, 5,
+
+def wasm_memcpylike_t : SDTypeProfile<0, 5,
```
[the remark this answers is not in the document]
Comment (sunfishcode) [reply llvm/lib/Target/WebAssembly/WebAssemblyInstrBulkMemory.td]: Fixed.

**Comment by aheejin** [comment]
The tests still don't seem to be passing.

File: llvm/lib/Target/WebAssembly/WebAssemblyInstrBulkMemory.td [hunk 4]
```diff
@@ -21,22 +21,37 @@ multiclass BULK_I<dag oops_r, dag iops_r, dag oops_s, dag iops_s,
...
-
-def wasm_memset_t : SDTypeProfile<0, 4,
+def wasm_memsetlike_t : SDTypeProfile<0, 4,
   [SDTCisInt<0>, SDTCisPtrTy<1>, SDTCisInt<2>, SDTCisInt<3>]
 >;
-def wasm_memset : SDNode<"WebAssemblyISD::MEMORY_FILL", wasm_memset_t,
+
+// memory.copy (may trap on empty ranges)
+def wasm_memory_copy : SDNode<"WebAssemblyISD::MEMORY_COPY", wasm_memcpylike_t,
+                              [SDNPHasChain, SDNPMayLoad, SDNPMayStore]>;
```
[the remark this answers is not in the document]
Comment (sunfishcode) [reply llvm/lib/Target/WebAssembly/WebAssemblyInstrBulkMemory.td]: Done.

**Review by aheejin — APPROVED** [review]

**Pull request merged**